### PR TITLE
consume errlen so we don't use garbage pass end of string

### PR DESCRIPTION
**Pull request by kahing**
counterpart of https://github.com/openresty/lua-nginx-module/pull/1165

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-resty-core project.

### CODE REVIEW

File: lib/resty/core/var.lua [hunk 1]
```diff
@@ -108,7 +109,7 @@ local function var_set(self, name, value)
     end
 
     if rc == -1 then  -- NGX_ERROR
-        return error(ffi_str(errbuf, errlen))
+       return error(ffi_str(errbuf, errlen[0]))
```
Comment (agentzh) [inline lib/resty/core/var.lua]: 3 space indentation?

**Comment by agentzh** [comment]
@kahing I wonder if it's possible to add a test case to the existing test suite to actually cover this fix?

File: lib/resty/core/var.lua (hunk 1 above)
Comment (kahing) [reply lib/resty/core/var.lua]: fixed

**Comment by kahing** [comment]
I considered that but I am not familiar with the test suite enough. This problem in theory only exhibits itself in the error log.

**Comment by agentzh** [comment]
@kahing The test cases are declarative and the test scaffold is explained in detail in this tutorial:

https://openresty.gitbooks.io/programming-openresty/content/testing/

And the test scaffold can surely check error logs, among many other things. Will you give it a shot?

**Comment by kahing** [comment]
honestly not sure if I have time for that, this internally is a very low priority fix.

File: lib/resty/core/var.lua [hunk 2]
```diff
@@ -108,7 +109,7 @@ local function var_set(self, name, value)
     end
 
     if rc == -1 then  -- NGX_ERROR
-        return error(ffi_str(errbuf, errlen))
+        return error(ffi_str(errbuf, errlen[0]))
     end
 end
```
Comment (agentzh) [inline lib/resty/core/var.lua]: This fix should be covered by a test case like below:

```
diff --git a/t/var.t b/t/var.t
index bdb0e59..7fc1880 100644
--- a/t/var.t
+++ b/t/var.t
@@ -215,3 +215,28 @@ qr/\[TRACE   \d+ content_by_lua\(nginx\.conf:\d+\):3 loop\]/
 --- no_error_log
 [error]
  -- NYI:
+
+
+
+=== TEST 8: error buffer overread
+--- http_config eval: $::HttpConfig
+--- config
+    location = /test {
+        content_by_lua_block {
+            local ok1, err1 = pcall(function () ngx.var.foo = 32; end)
+            local ok2, err2 = pcall(function () ngx.var.server_port = 32; end)
+            assert(not ok1)
+            ngx.say(err1)
+            assert(not ok2)
+            ngx.say(err2)
+        }
+    }
+--- request
+GET /test
+--- response_body
+content_by_lua(nginx.conf:62):2: variable "foo" not found for writing; maybe it is a built-in variable that is not changeable or you forgot to use "set $foo '';" in the config file to define it first
+content_by_lua(nginx.conf:62):3: variable "server_port" not changeable
+--- no_error_log
+[error]
+[alert]
+ -- NYI:
```

**Comment by agentzh** [comment]
@kahing Committed with commit c0dcdc657e4 with various fixes suggested above. Thanks for your contribution.

**Comment by kahing** [comment]
thanks for pushing this over the finish line!

**Pull request closed without merging**